keps/sig-node/3063-dynamic-resource-allocation/README.md

```diff
@@ -0,0 +1,2052 @@
+pods to the node that normally would run elsewhere. With dynamic resource
+allocation, such an attack is still possible, but the attack code would have to
+be different for each resource driver because all of them will use their own,
+custom approach for reporting resource availability.
+
+The security of those custom approaches is the responsibility of the resource
+driver vendor. Solutions like Akri which establish their own control plane and
+then communicate with Kubernetes through the device plugin API already need to
+address this.
+
+Aside from security implications, usability and usefulness of dynamic resource
+allocation also may turn out to be insufficient. Some risks are:
+
+- Slower pod scheduling due to the interaction with resource drivers.
+
+- Additional complexity when describing pod requirements because
+  separate objects must be created for the parameters.
+
+- Network-attached resources may have additional constraints that are not
+  captured yet (like limited number of nodes that they can be attached to).
+
+- Cluster autoscaling will not work as expected unless the autoscaler and
+  resource drivers get extended to support it.
+
+All of these risks will have to be evaluated by gathering feedback from users
+and resource driver developers.
+
+## Design Details
+
+
+### Implementation
+
+![components](./components.png)
+
+Several components must be implemented or modified in Kubernetes:
+- The new API must be added to kube-apiserver.
+- A new controller in kube-controller-manager which creates 
+  ResourceClaims from Pod ResourceClaimTemplates, similar to
+  https://github.com/kubernetes/kubernetes/tree/master/pkg/controller/volume/ephemeral
+- A kube-scheduler plugin must detect Pods which reference a
+  ResourceClaim (directly or through a template) and ensure that the
+  resource is allocated before the Pod gets scheduled, similar to
+  https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/volume/scheduling/scheduler_binder.go
+- Kubelet must be extended to retrieve information from ResourceClaims
+  and to call a resource kubelet plugin. That plugin returns CDI device ID(s)
+  which then must be passed to the container runtime.
+
+For a resource driver the following components are needed:
+- *Resource driver controller*: a central component which handles resource allocation
+  by watching ResourceClaims and updating their status once it is done with
+  allocation. It may run inside the cluster or outside of it. The only
+  hard requirement is that it can connect to the API server. Optionally,
+  it may also be configured as a [special scheduler extender](#kube-scheduler).
+- *Resource kubelet plugin*: a component which cooperates with kubelet to prepare
+  the usage of the resource on a node.
+
+The utility library will be developed outside of Kubernetes and does not have
+to be used by drivers, therefore it is not described further in this KEP.
+
+### Resource allocation flow
```

Comment:
> The current flow has 5 API calls (as opposed to 1 for regular pods):

While implementing the prototype, I found that this can be optimized. I should have updated the flow accordingly, but missed it - my bad, sorry for that. I'll try to replace it with pseudo code, if I find that this becomes clearer than the diagram, otherwise update the diagram.

The optimization is that the scheduler can combine "kube-scheduler writes potentialNodes" with "kube-scheduler writes selected node". That works because for a new claim `UnsuitableNodes` is empty and the scheduler can and will proceed with selecting a node. The semantic is intentionally inverted so that it doesn't need to wait for a positive node list from the driver.

The controller sets the finalizer, allocates, and updates the claim with the result.

The scheduler updates ReservedFor and sets Pod.spec.nodeName.

> What if kube-scheduler sends a sorted list of nodes instead (after scoring). Then the controller directly reserves the ResourceClaim for the first node that can have it.

What if the pod uses two resources from different drivers? When both pick a node from the list independently, it might not be the same node.

I also don't see where a claim gets reserved for the Pod. A claim might get reused, so "allocated" and "reserved" must be different aspects of it.

Perhaps we can optimize the flow by adding a field "SelectedPod" that gets set together with "SelectedNode": when the driver is done with the allocation, it can add that pod as user in the same claim update that also sets the allocation result.